lib/credit-processor/src/main/scala/com/gu/creditprocessor/Processor.scala

```diff
@@ -106,17 +108,37 @@ object Processor {
         val creditRequests = creditRequestsFromSalesforce.distinct
         val alreadyActionedCredits = creditRequestsFromSalesforce.flatMap(_.chargeCode).distinct
         logger.info(s"Processing ${creditRequests.length} credits in Zuora ...")
-        val allZuoraCreditResponses = creditRequests.map(
-          addCreditToSubscription(
-            creditProduct,
-            getSubscription,
-            getAccount,
-            updateToApply,
-            updateSubscription,
-            resultOfZuoraCreditAdd,
-            getNextInvoiceDate,
-          ),
-        )
+
+        val bySub =
+          creditRequests
+            .groupBy(_.subscriptionName)
+            .values
+            .toList
+            .par
+
+        val requestConcurrency = 20
+        /*https://developer.zuora.com/docs/guides/rate-limits/#concurrent-request-limits
+        Zuora supports up to 40 concurrent requests until migration to orders API which supports 200
+         */
+        val forkJoinPool = new java.util.concurrent.ForkJoinPool(requestConcurrency)
+
+        bySub.tasksupport = new ForkJoinTaskSupport(forkJoinPool)
+        val allZuoraCreditResponses = bySub.flatMap { requests =>
+          requests.map(
+            addCreditToSubscription(
+              creditProduct,
+              getSubscription,
+              getAccount,
+              updateToApply,
+              updateSubscription,
+              resultOfZuoraCreditAdd,
+              getNextInvoiceDate,
+            ),
+          )
+        }.toList
```

Comment:
not sure if we can change to for comprehension without making it do the contents of each batch in parallel?
also a few naming update suggestions
```suggestion
        val allZuoraCreditResponses = (for {
          batch <- creditRequestBatches
          request <- batch
        } yield addCreditToSubscription(
            creditProduct,
            getSubscription,
            getAccount,
            updateToApply,
            updateSubscription,
            resultOfZuoraCreditAdd,
            getNextInvoiceDate,
          )(request)
        ).toList
```